File: sdk/python/examples/langgraph/11_customer_support.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Customer Support Router — create_agent with category-specific response tools.

Demonstrates:
    - Using create_agent with a system prompt that directs routing behaviour
    - Tools for billing, technical, and general support handlers
    - The LLM classifies and routes to the right tool server-side

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def handle_billing(user_message: str) -> str:
    """Handle a billing-related customer inquiry.

    Provides empathetic billing support, explains payment options, and offers
    to review the account. Use this for payment, invoice, charge, or refund questions.

    Args:
        user_message: The customer's billing question or complaint.
    """
    return (
        f"Hello! Thank you for contacting our support team. I'm here to help you today.\n\n"
        f"I understand you have a billing concern. I sincerely apologize for any confusion "
        f"with your account. Our billing team can review your charges and process any eligible "
        f"refunds within 3-5 business days. You can also update your payment method anytime "
        f"in your account settings. Is there anything specific about your invoice I can clarify?"
    )


@tool
def handle_technical(user_message: str) -> str:
    """Handle a technical support inquiry with step-by-step troubleshooting guidance.

    Use this for software bugs, connectivity issues, configuration problems,
    or any technical malfunction.

    Args:
        user_message: The customer's technical issue or error description.
    """
    return (
        f"Hello! Thank you for contacting our support team. I'm here to help you today.\n\n"
        f"I can see you're experiencing a technical issue. Let's work through this together:\n"
        f"1. First, try clearing your browser cache and cookies, then reload the page.\n"
        f"2. If the issue persists, log out and log back in to refresh your session.\n"
        f"3. Check our status page at status.example.com for any known outages.\n"
        f"4. If none of these work, please share your error message or screenshot "
        f"and I'll escalate to our engineering team immediately."
    )


@tool
def handle_general(user_message: str) -> str:
    """Handle a general customer inquiry with friendly, helpful guidance.

    Use this for account questions, feature inquiries, onboarding help,
    or any question that doesn't fit billing or technical categories.

    Args:
        user_message: The customer's general question or request.
    """
    return (
        f"Hello! Thank you for contacting our support team. I'm here to help you today.\n\n"
        f"Great question! I'm happy to help you with that. Our platform offers a wide range "
        f"of features designed to make your experience seamless. If you need help with account "
        f"settings, feature setup, or have any other questions, please don't hesitate to ask. "
        f"I'm here to assist! Is there anything else I can help you with today?"
    )


CS_SYSTEM = """You are a friendly and empathetic customer service representative.

When a customer contacts you:
1. Classify their inquiry as billing, technical, or general
2. Call the matching handler tool:
   - Payment/invoice/charge/refund issues → handle_billing
   - Software bugs/errors/connectivity issues → handle_technical
   - Everything else → handle_general
3. Return the handler's response directly to the customer.
"""

graph = create_agent(
    llm,
    tools=[handle_billing, handle_technical, handle_general],
    name="customer_support",
    system_prompt=CS_SYSTEM,
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(
            graph,
            "I was charged twice for my subscription this month and need a refund.",
        )
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/11-customer-support.ts
/**
 * Customer Support Router -- StateGraph with greet -> classify -> route -> respond.
 *
 * Demonstrates:
 *   - Multi-node StateGraph with conditional branching
 *   - Classifying user intent and routing to specialized handlers
 *   - Billing, technical, and general support branches
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// State schema
// ---------------------------------------------------------------------------
const SupportState = Annotation.Root({
  user_message: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  greeting: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  category: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  response: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type State = typeof SupportState.State;

const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function greet(_state: State): Partial<State> {
  return {
    greeting:
      'Hello! Thank you for contacting our support team. ' +
      "I'm here to help you today.",
  };
}

async function classify(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      "Classify the customer message into exactly one category: " +
        "'billing', 'technical', or 'general'. " +
        "Return only the single category word.",
    ),
    new HumanMessage(state.user_message),
  ]);
  let category = (response.content as string).trim().toLowerCase();
  if (!['billing', 'technical', 'general'].includes(category)) {
    category = 'general';
  }
  return { category };
}

function routeCategory(state: State): string {
  return state.category;
}

async function handleBilling(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'You are a billing specialist. The customer has a billing question. ' +
        'Be empathetic, offer to review their account, and explain payment options clearly.',
    ),
    new HumanMessage(state.user_message),
  ]);
  return { response: `${state.greeting}\n\n${response.content}` };
}

async function handleTechnical(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'You are a technical support engineer. The customer has a technical issue. ' +
        'Provide step-by-step troubleshooting guidance. Be clear and concise.',
    ),
    new HumanMessage(state.user_message),
  ]);
  return { response: `${state.greeting}\n\n${response.content}` };
}

async function handleGeneral(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'You are a helpful customer service agent handling general inquiries. ' +
        'Be friendly, informative, and direct. Offer additional help at the end.',
    ),
    new HumanMessage(state.user_message),
  ]);
  return { response: `${state.greeting}\n\n${response.content}` };
}

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const builder = new StateGraph(SupportState);
builder.addNode('greet', greet);
builder.addNode('classify', classify);
builder.addNode('billing', handleBilling);
builder.addNode('technical', handleTechnical);
builder.addNode('general', handleGeneral);

builder.addEdge(START, 'greet');
builder.addEdge('greet', 'classify');
builder.addConditionalEdges('classify', routeCategory, {
  billing: 'billing',
  technical: 'technical',
  general: 'general',
});
builder.addEdge('billing', END);
builder.addEdge('technical', END);
builder.addEdge('general', END);

const graph = builder.compile();

// Add agentspan metadata for extraction
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools: [],
  framework: 'langgraph',
};

const PROMPT =
  'I was charged twice for my subscription this month and need a refund.';

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    // Deploy to server. CLI alternative (recommended for CI/CD):
    //   agentspan deploy <module>
    // await runtime.deploy(graph);
    // await runtime.serve(graph);
    // Direct run for local development:
    const result = await runtime.run(graph, PROMPT);
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('11-customer-support.ts') || process.argv[1]?.endsWith('11-customer-support.js')) {
  main().catch(console.error);
}
